File: 206_Reverse_Linked_list.java
class Solution {

    public ListNode reverseList(ListNode head) {

        /* Sanity check */
        if(head == null || head.next == null)
            return head;

        ListNode p1 = head;
        ListNode p2 = head.next;
        ListNode p3 = head.next.next;
        p1.next = null;

        while(true) {
            p2.next = p1;

            /* Termination */
            if(p3 == null)
                return p2;

            p1 = p2;
            p2 = p3;
            p3 = p3.next;
        }
    }

}

File: 206_Reverse_Linked_List.cpp
//
//  206. Reverse Linked List.h
//  Project 1
//
//  Created by Jonny Kong on 2/22/16.
//  Copyright © 2016 Jonny Kong. All rights reserved.
//

ListNode* reverseList(ListNode* head) {
  ListNode *p = head, *q = head, *address = head;

  if (!head)
    return NULL;

  else if (head->next == NULL)
    return head;

  q = q->next;

  while (q) {
    address = q->next;

    q->next = p;

    p = q;

    q = address;
  }

  head->next = NULL;

  return p;
}